### PR TITLE
SMART Metadata URLs

### PR DESCRIPTION
**Pull request by proteusvacuum**
## Summary
Builds out the metadata urls required for SMART discovery (http://hl7.org/fhir/smart-app-launch/conformance/index.html#using-well-known and https://www.hl7.org/fhir/capabilitystatement.html)

Base branch is https://github.com/dimagi/commcare-hq/pull/29210 to prevent wicked merge conflicts later. 
I filled out as little of these two endpoints as possible for the POC. I'd be interested in feedback on whether I should do this more formally with https://github.com/smart-on-fhir/client-py or if there is a better way of building out these super nested but sparse JSON docs. 

This is a spin out from the larger SMART-on-FHIR work. The few commented lines of code come from there, and will be replaced later. 

## Feature Flag

FHIR_INTEGRATION

## Safety Assurance

- [x] Risk label is set correctly
- [x] All migrations are backwards compatible and won't block deploy
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change
- [x] If QA is part of the safety story, the "Awaiting QA" label is used
- [x] I have confidence that this PR will not introduce a regression for the reasons below
Fr/smart metadata urls
### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations

### CODE REVIEW

**Comment by mkangia** [comment]
> I'd be interested in feedback on whether I should do this more formally with https://github.com/smart-on-fhir/client-py or if there is a better way of building out these super nested but sparse JSON docs.

We have not used any external library yet yes? So i think its okay to just build this ourselves, at least for the POC bit.

File: corehq/motech/fhir/models.py [hunk 1]
```diff
@@ -334,3 +337,103 @@ def deepmerge(a, b):
...
+    )
+    registration_endpoint = StringProperty(exclude_if_none=True)
+    scopes_supported = ListProperty(StringProperty, exclude_if_none=True)
+    response_types_supported = StringProperty(exclude_if_none=True)
+    management_endpoint = StringProperty(exclude_if_none=True)
+    revokation_endpoint = StringProperty(exclude_if_none=True)
+    capabilities = ListProperty(StringProperty, exclude_if_none=True)
+
+
+class _Coding(JsonObject):
```
Comment (mkangia) [inline corehq/motech/fhir/models.py:360]: nit: I think a better order for these classes would be
```
_MetadataRest
_RestSecurity
_CodableConcept
_Coding
_Extension
_ExtensionURI
```
the current order made a bit hard to follow the code for review.

(I keep forgetting the name for this flow of code, something we read as a part of clean code book)

File: corehq/motech/fhir/models.py [hunk 2]
```diff
@@ -334,3 +337,103 @@ def deepmerge(a, b):
...
+    """
+    The /metadata response for a domain which contains the FHIR CapabilityStatement
+    http://hl7.org/fhir/capabilitystatement.html
+    """
+    format = ListProperty(StringProperty(), required=True, default=['json'])
+    status = StringProperty(required=True, default='active')
+    kind = StringProperty(required=True, default='instance')
+    fhirVersion = StringProperty(required=True, default=FHIR_VERSION_4_0_1)
+    rest = ListProperty(_MetadataRest, required=True, default=[_MetadataRest()])
+    date = StringProperty(required=True, default='2021-03-23')
```
Comment (mkangia) [inline corehq/motech/fhir/models.py:406]: why this date?

**Review by mkangia — COMMENTED** [review]
when i was thinking of this initially, I thought these responses could just be kept as a template in a file/ or in code itself and then filled with values when called for. 
We are not supporting much anyway as of now.
Do we need to add models? Or we need that for Smart or anything else?

File: corehq/motech/fhir/models.py (hunk 2 above)
Comment (proteusvacuum) [reply corehq/motech/fhir/models.py:406]: This is the "date (and optionally time) when the capability statement was published. The date must change when the business version changes and it must change if the status code changes. In addition, it should change when the substantive content of the capability statement changes." https://www.hl7.org/fhir/capabilitystatement-definitions.html#CapabilityStatement.date
I will make this clearer in the code.

File: corehq/motech/fhir/models.py (hunk 1 above)
Comment (proteusvacuum) [reply corehq/motech/fhir/models.py:360]: Yeah I agree that this is a bit confusing, but unfortunately these have to remain in this order because the dependencies flow downwards. (`_CodableConcept` needs `_Coding` to be defined first, etc)

File: corehq/motech/fhir/models.py (hunk 2 above)
Comment (proteusvacuum) [reply corehq/motech/fhir/models.py]: https://github.com/dimagi/commcare-hq/pull/29453/commits/b5e2b13297b36f1fcdaf4ec9be4ec714a6a44efc

**Comment by proteusvacuum** [comment]
@mkangia thanks. I was definitely over thinking this!

File: corehq/motech/fhir/utils.py [hunk 3]
```diff
@@ -53,3 +58,20 @@ def update_fhir_resource_property(case_property, fhir_resource_type, fhir_resour
...
         fhir_resource_prop.jsonpath = fhir_resource_prop_path
         fhir_resource_prop.save()
+
+
+def build_capability_statement(domain):
+    """
+    Builds the FHIR capability statement including the OAuth URL extensions for SMART
+    https://hl7.org/fhir/smart-app-launch/conformance/index.html
+    """
+    with open(os.path.join(os.path.dirname(__file__), 'data', 'capability_statement.json'), encoding='utf-8') as f:
```
Comment (mkangia) [inline corehq/motech/fhir/utils.py:68]: I think we can just cache this template in memory to avoid the IO each time this is called?

```
@memoized
def load_capability_statement():
    return open(os.path.join(os.path.dirname(__file__), 'data', 'capability_statement.json'), encoding='utf-8').read()
```

File: corehq/motech/fhir/data/capability_statement.json [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+        }
+    ],
+    "format": [
+        "json"
+    ],
+    "status": "active",
+    "kind": "instance",
+    "fhirVersion": "__FHIR_VERSION__",
+    "date": "__PUBLISHED_DATE__"
+}
```
Comment (kaapstorm) [inline corehq/motech/fhir/data/capability_statement.json:41]: omg yes! Thank you!

Every time we statically define something that could just be a dictionary, a Python fairy dies, and a Java demon does that "MWHAHAHAHA" laugh from the pit of Python fairy hell, knowing that every time a developer needs to do anything with what could just have been a dictionary, they first have to negotiate with nested classes.

File: corehq/motech/fhir/utils.py [hunk 5]
```diff
@@ -53,3 +58,20 @@ def update_fhir_resource_property(case_property, fhir_resource_type, fhir_resour
...
+        data = (
+            f.read()
+            .replace('__TOKEN_URL__', absolute_reverse("oauth2_provider:token"))
+            .replace('__AUTHORIZE_URL__', absolute_reverse("oauth2_provider:authorize"))
+            .replace('__FHIR_VERSION__', FHIR_VERSION_4_0_1)
+            .replace('__PUBLISHED_DATE__', CAPABILITY_STATEMENT_PUBLISHED_DATE)
+        )
+        statement = json.loads(data)
```
Comment (kaapstorm) [inline corehq/motech/fhir/utils.py:76]: Instead of storing this in a file, or caching it in memory, we could just define it as a dictionary, and update it using the `deepmerge()` function that I killed (with profound sadness) in [this commit here](https://github.com/dimagi/commcare-hq/pull/29456/commits/cc279b3393b897c97f676f0559ea880c8085ff47).

e.g.
```python
    statement = deepmerge(statement, {
        'rest': [{'security': {'extension': {'extension': [
            {'valueUri': absolute_reverse("oauth2_provider:token")},
            {'valueUri': absolute_reverse("oauth2_provider:authorize")},
        ]}}}],
        'fhirVersion': FHIR_VERSION_4_0_1,
        'date': CAPABILITY_STATEMENT_PUBLISHED_DATE,
    })
```

**Review by kaapstorm — APPROVED** [review]

File: corehq/motech/fhir/utils.py (hunk 3 above)
Comment (proteusvacuum) [reply corehq/motech/fhir/utils.py]: https://github.com/dimagi/commcare-hq/pull/29453/commits/6ccce12c009e78aa868273b1bcb4b02ce34f7e88

File: corehq/motech/fhir/utils.py [hunk 6]
```diff
@@ -53,3 +56,51 @@ def update_fhir_resource_property(case_property, fhir_resource_type, fhir_resour
                                                       resource_type=fhir_resource_type)
         fhir_resource_prop.jsonpath = fhir_resource_prop_path
         fhir_resource_prop.save()
+
+
+def build_capability_statement(domain, fhir_version):
```
Comment (proteusvacuum) [inline corehq/motech/fhir/utils.py:61]: I am going to use the `domain` param later as the token and authorize urls are going to be domain specific.

**Review by kaapstorm — APPROVED** [review]

File: corehq/motech/fhir/data/capability_statement.json (hunk 4 above)
Comment (proteusvacuum) [reply corehq/motech/fhir/data/capability_statement.json]: Yeah, I don't know what I was thinking.

Comment (kaapstorm) [reply corehq/motech/fhir/data/capability_statement.json]: Oh, as you know, there is a lot of precedent in our codebase, and I've done the same myself. (In a pre-Dimagi life I spent a year building a platform that (among other things) inspected remote APIs and used Python metaclasses to generate classes to represent API documents/resources. Way down the line I realized that I could have avoided all the metaclass work by just using dictionaries, and, if they needed to be validated, checking them against a generated schema -- i.e. what we are doing with FHIR resources.)
